packages/images/src/service.js

```diff
@@ -0,0 +1,199 @@
+import multer from 'multer';
+import sharp from 'sharp';
+import crypto from 'crypto';
+import uuid from 'uuid/v4';
+
+// http://localhost:4008/image/5d5a561875a3821da888fe2570d910ea.jpeg
+// http://localhost:4008/image/5d5a561875a3821da888fe2570d910ea.png
+// http://localhost:4008/image/5d5a561875a3821da888fe2570d910ea.jpeg?width=200&height=100&fit=cover
+// http://localhost:4008/image/5d5a561875a3821da888fe2570d910ea.jpeg?width=200&height=100&fit=contain
+// http://localhost:4008/image/5d5a561875a3821da888fe2570d910ea.jpeg?width=200&height=100&fit=cover
+// http://localhost:4008/image/5d5a561875a3821da888fe2570d910ea.jpeg?width=200&fit=cover
+// http://localhost:4008/image/5d5a561875a3821da888fe2570d910ea.jpeg?height=200&fit=cover
+
+const VALID_FIT_TYPES = ['cover', 'contain', 'fill', 'inside', 'outside'];
+
+class DataStore {
+  constructor({ path }) {
+    this.jsonPath = `${path}/data.json`;
+    try {
+      this.data = fs.readJsonSync(this.jsonPath);
+    } catch (err) {
+      if (err.code === 'ENOENT') {
+        this.data = {};
+      } else {
+        throw err;
+      }
+    }
+  }
+  storeImage(id, meta) {
+    this.data[id] = meta;
+    fs.outputJSONSync(this.jsonPath, this.data);
+  }
+  getImage(id) {
+    return this.data[id];
+  }
+}
+
+const processResizeOptions = ({ width, height, fit = 'cover' }) => {
+  if (!width && !height) {
+    return null;
+  }
+  if (!VALID_FIT_TYPES.includes(fit)) {
+    return {
+      error: `Invalid fit option; can be one of ${VALID_FIT_TYPES.join(', ')}.`,
+    };
+  }
+  const widthInt = parseInt(width, 10);
+  const heightInt = parseInt(height, 10);
+
+  if (width && (isNaN(widthInt) || widthInt < 1)) {
+    return {
+      error: `Invalid width. Please provide a positive integer`,
+    };
+  }
+
+  if (height && (isNaN(heightInt) || heightInt < 1)) {
+    return {
+      error: `Invalid height. Please provide a positive integer`,
+    };
+  }
+
+  return { width: widthInt || undefined, height: heightInt || undefined, fit };
+};
+
+export class ImageService {
+  constructor({ port = 4001, protocol = 'http', host = 'localhost', path = './images' }) {
+    this.path = nodePath.resolve(path);
+    this.protocol = protocol;
+    this.port = port;
+    this.host = host;
+    this.database = new DataStore({ path: this.path });
+    this.sourcePath = nodePath.join(this.path, 'source');
+    this.transformsPath = nodePath.join(this.path, 'transforms');
+    fs.ensureDirSync(this.sourcePath);
+    fs.ensureDirSync(this.transformsPath);
+
+    const upload = multer({ dest: this.sourcePath });
+
+    this.app = express();
+
+    this.app.post('/upload', upload.single('image'), async (req, res) => {
+      const { originalname, path } = req.file;
+
+      res.status(201).json(this.storeImageInDatabase({ path, originalname }));
+    });
+
+    this.app.get('/image/:filename', async (req, res) => {
+      const { filename } = req.params;
+      const [id, format] = filename.split('.');
```

Comment:
I haven't adequately looked into how files are stored, but is it possible to have a file with `.`s in its name?

Reply:
It's fine (for now)

We control the filenames that are created, so until they are made customisable on upload, this is not going to create a problem